### ISSUE TITLE
Modified data-params attribute doesn't get submitted

### ISSUE THREAD

**Issue by etagwerker** [rails/jquery-ujs#505]
Hello, 

First of all, thanks for all your hard work! I've been using jquery-ujs for years without a problem.

Secondly, I think I've found an issue or I'm not using jquery-ujs the way it's supposed to. 

Here is the code that I'm using:

```
link_to "Send Terms",
  email_terms_school_students_path(@current_school),
  method: :post,
  remote: true,
  data: { confirm: "Are you sure?" }
```

I have a table and I have a piece of JS code that looks like this:

```
var updateDataAttributes = function() {
  var student_ids = $('input[name^=student_ids]:checked').map(function() {
    return $(this).attr('value');
  }).get();
  var student_ids_tuple = JSON.stringify({student_ids: student_ids});

  $('#bulk_actions a').attr('data-params', student_ids_tuple);
}
```

The first time that I set the `data-params` attribute, I see my page make the right request with this data:

```
student_ids%5B%5D=74&student_ids%5B%5D=50
```

That works perfectly. 

But then I select 2 more items in my table (for a total of 4 selected) and I see that the `data-params` attribute gets updated with 2 more ids. I now have 4 ids in the `data-params` attribute.

Yet when I debug the code, I see that jQuery is sending only the first two ids: 

```
student_ids%5B%5D=74&student_ids%5B%5D=50
```

Is this getting cached somewhere? I would expect it to send 4 ids instead of 2.

I found a workaround for this problem but it's a bit hacky and it does the same thing I'm expecting jQuery UJS to do.

This is my workaround: 

```
  $('#bulk_actions a').on('ajax:beforeSend', function(evt, xhr, settings) {
    var student_ids = $('input[name^=student_ids]:checked').map(function() {
      return $(this).attr('value');
    }).get();

    var data = "";

    student_ids.forEach(function(id) {
      data += "student_ids%5B%5D=" + id + "&";
    });

    settings.data = data;
  });
```

Maybe I'm using it the wrong way? Is it supported to edit the `data-params` attribute dynamically with jQuery? 

Please let me know.

Thanks!

**Comment by jeremy** [comment]
Sounds like you're on the right track. Have you traced it with a debugger to see exactly what's happening?

**Comment by etagwerker** [comment]
@Jeremy Yes, I've debugged it and it's happening with links like these one: 

```
<a data-confirm="Are you sure?" data-remote="true" rel="nofollow" data-method="post" href="/schools/1/students/accept_terms" data-params="{&quot;record_ids&quot;:[&quot;267&quot;,&quot;290&quot;,&quot;284&quot;,&quot;288&quot;,&quot;281&quot;,&quot;270&quot;,&quot;272&quot;,&quot;264&quot;]}">Mark Permission as Accepted</a>
```

It's calling `handleRemote` and within it there is an odd behavior. 

It's executing this path: https://github.com/rails/jquery-ujs/blob/master/src/rails.js#L138-L140

When I debug it in there, this is what I see:

```
// this is correct
> element.attr('data-params') 
"{"record_ids":["267","290","284","288","281","270","272","264"]}"

// but this is wrong
> element.data('params')
{record_ids: Array(2)}
record_ids
:
(2) ["267", "290"]
```

Unfortunately it seems to be an issue with jQuery. I would expect `element.data('params')` to remove the right `Array` but it's returning an old one (the first one)

I found this occurrence in another project: https://github.com/meteor/blaze/issues/155#issuecomment-260144553

And I found this answer in SO: https://stackoverflow.com/a/8708345/615138

Anyway, TL;DR, `element.data` is not an _accessor_ for `data-` attributes: http://api.jquery.com/data/

That's why modifying any of the `data-` attributes in the DOM won't work with jQuery UJS because it relies on `element.data` and not `element.attr`

**Comment by jeremy** [comment]
Ahh right, classic! Nice catch. This is the "jQuery way," so we'll continue to roll with it.
